pkg/operator/controller/extension/required/virtual/reconciler.go

```diff
@@ -0,0 +1,91 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	"github.com/gardener/gardener/pkg/operator/apis/config"
+)
+
+// Reconciler reconciles Extensions to determine their required state.
+type Reconciler struct {
+	Config        config.ExtensionRequiredVirtualControllerConfiguration
+	RuntimeClient client.Client
+	VirtualClient client.Client
+
+	clock clock.Clock
+}
+
+// Reconcile processes the given extension object in the request.
+// It lists required ControllerInstallations to ascertain if the extension is required in seed clusters.
+// At the end, the RequiredVirtual condition is updated for the extension.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	extension := &operatorv1alpha1.Extension{}
+	if err := r.RuntimeClient.Get(ctx, request.NamespacedName, extension); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	controllerInstallation := &gardencorev1beta1.ControllerInstallationList{}
+	if err := r.VirtualClient.List(ctx, controllerInstallation, client.MatchingFields{gardencore.RegistrationRefName: extension.Name}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error listing controllerinstallations: %w", err)
+	}
+
+	requiredSeeds := sets.New[string]()
+	for _, ctrlInstallation := range controllerInstallation.Items {
+		requiredCondition := v1beta1helper.GetCondition(ctrlInstallation.Status.Conditions, gardencorev1beta1.ControllerInstallationRequired)
+		if requiredCondition != nil && requiredCondition.Status == gardencorev1beta1.ConditionTrue {
+			requiredSeeds.Insert(ctrlInstallation.Spec.SeedRef.Name)
+		}
+	}
+
+	if err := r.updateCondition(ctx, log, extension, requiredSeeds); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not update extension status: %w", err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) updateCondition(ctx context.Context, log logr.Logger, extension *operatorv1alpha1.Extension, seedNames sets.Set[string]) error {
+	requiredCondition := v1beta1helper.GetOrInitConditionWithClock(r.clock, extension.Status.Conditions, operatorv1alpha1.ExtensionRequiredVirtual)
+
+	if seedNames.Len() > 0 {
+		sortedSeedNames := slices.Sorted(slices.Values(seedNames.UnsortedList()))
+		requiredCondition = v1beta1helper.UpdatedConditionWithClock(r.clock, requiredCondition, gardencorev1beta1.ConditionTrue, "RequiredControllerInstallation", fmt.Sprintf("Extension has required ControllerInstallations for seed clusters %s", sortedSeedNames))
```

Comment:
Should we really out all seed clusters in the condition message? I guess this can get quite long

Reply:
I was also going back and forth when adding this message and decided to remove it again with your backing opinion 😄